lib/galaxy/objectstore/irods.py

```diff
@@ -419,18 +419,26 @@ def _push_to_irods(self, rel_path, source_file=None, from_string=None):
                 log.debug("Wanted to push file '%s' to iRODS collection '%s' but its size is 0; skipping.", source_file, rel_path)
                 return True
 
+            # Create sub-collection first
+            self.session.collections.create(collection_path, recurse=True)
+
+            # Create data object
+            data_obj = self.session.data_objects.create(data_object_path, self.resource, **options)
+
             if from_string:
-                data_obj = self.session.data_objects.create(data_object_path, self.resource, **options)
+                # Save 'from_string' as a file
                 with data_obj.open('w') as data_obj_fp:
                     data_obj_fp.write(from_string)
+
+                # Add file containing 'from_string' to the irods collection, since
+                # put() expects a file as input. Get file name from data object's 'desc' field
+                self.session.data_objects.put(data_obj.desc, collection_path + "/", **options)
```

Comment:
Must use put() to make code indifferent to storage type (S3, disk, etc). put() requires a file as input. We can use the data object file saved above (file path given by data object's 'desc' field)